scene/resources/paragraph_layout.cpp

```diff
@@ -0,0 +1,1655 @@
+
+	_runs.clear();
+	_runs_visual_to_logical_map.clear();
+
+	_justification_opportunies.clear();
+	_break_opportunies.clear();
+
+	_dirty_text_tokens = true;
+	_dirty_shaped_tokens = true;
+	_dirty_line_breaks = true;
+	_dirty_line_layout = true;
+};
+
+void ParagraphLayout::add_text(const String &p_string, Ref<Font> p_font, const String &p_lang, const String &p_features) {
+
+	UErrorCode err = U_ZERO_ERROR;
+
+	if (p_font.is_null()) return;
+
+	if (p_string.length() > 0) {
+
+		if ((_style_runs.size() > 0) && (_style_runs[_style_runs.size() - 1].font == p_font) && (_style_runs[_style_runs.size() - 1].language == p_lang) && (_style_runs[_style_runs.size() - 1].features == p_features)) {
+			_style_runs[_style_runs.size() - 1].end = _utf16_text.size() + p_string.length();
+		} else {
+			_style_runs.push_back(RunLimit(p_font, p_lang, p_features, _utf16_text.size(), _utf16_text.size() + p_string.length()));
+		}
+
+		if (sizeof(CharType) == 4) {
+			//:'‑( wchar_t why???
+			int32_t utf16_length;
+			size_t old_length = _utf16_text.size();
+			u_strFromUTF32(NULL, 0, &utf16_length, (const UChar32 *)p_string.ptr(), p_string.length(), &err);
+			_utf16_text.resize(_utf16_text.size() + utf16_length);
+			err = U_ZERO_ERROR;
+			u_strFromUTF32((UChar *)&_utf16_text[old_length], utf16_length, &utf16_length, (const UChar32 *)p_string.ptr(), p_string.length(), &err);
+		} else {
+			size_t old_length = _utf16_text.size();
+			_utf16_text.resize(_utf16_text.size() + p_string.length());
+			memcpy((UChar *)&_utf16_text[old_length], p_string.ptr(), p_string.length() * sizeof(UChar));
+		}
+
+		//TODO add validation
+
+		_dirty_text_tokens = true;
+		_dirty_shaped_tokens = true;
+		_dirty_line_breaks = true;
+		_dirty_line_layout = true;
+	}
+}
+
+void ParagraphLayout::add_stt_token(int p_start, int p_end) {
+
+	_text_tokens.push_back(StructuredTextToken(p_start, p_end));
+
+	_dirty_text_tokens = true;
+	_dirty_shaped_tokens = true;
+	_dirty_line_breaks = true;
+	_dirty_line_layout = true;
+}
+
+_FORCE_INLINE_ bool ParagraphLayout::_arabic_can_connect_to_left(UChar p_ch, UChar p_prev_ch) const {
+
+	//characters that are connectable to the left
+	bool ok = (0x0628 == p_prev_ch) || (0x062A <= p_prev_ch && p_prev_ch <= 0x062E) || (0x0633 <= p_prev_ch && p_prev_ch <= 0x0647) || (0x0649 == p_prev_ch) || (0x064A == p_prev_ch) || (0x0678 <= p_prev_ch && p_prev_ch <= 0x0687) || (0x069A <= p_prev_ch && p_prev_ch <= 0x06C1) || (0x06C3 <= p_prev_ch && p_prev_ch <= 0x06D3) || (0x06FA <= p_prev_ch && p_prev_ch <= 0x06FC);
+
+	//ignore Lam + Alef ligature
+	if (ok)
+		ok = !((p_prev_ch == 0x0644 || (p_prev_ch >= 0x06B5 && p_prev_ch <= 0x06B8)) && (p_ch == 0x0622 || p_ch == 0x0623 || p_ch == 0x0625 || p_ch == 0x0627 || p_ch == 0x0622 || p_ch == 0x0671 || p_ch == 0x0672 || p_ch == 0x0673 || p_ch == 0x0675));
+	return ok;
+}
+
+_FORCE_INLINE_ void ParagraphLayout::_generate_kashida_justification_opportunies(size_t p_start, size_t p_end) const {
```

Comment:
To support more languages using the Arabic script, I suggest instead of hard-coding character groups, I suggest using `UCHAR_JOINING_GROUP` character property. Here is how I’m doing it in LibreOffice:
https://cgit.freedesktop.org/libreoffice/core/tree/sw/source/core/text/porlay.cxx#n63